### ISSUE TITLE
Incorporate Plugin Repository functionality into Geeklog and Finish Repository Manager Plugin

### ISSUE THREAD

**Issue by eSilverStrike** [Geeklog-Core/geeklog#822]
Should we or not?  I hate to see the code go to waste but it was developed in 2009 so there may be a lot of changes and testing we will have to do.

http://wiki.geeklog.net/SoC_plugin_repository

https://github.com/Geeklog-GSoC/Repository-Core (Based on Geeklog 1.6.0 I believe)
https://github.com/Geeklog-GSoC/Repository-Plugin

**Comment by eSilverStrike** [comment]
@mystralkk Well I have put a number of hours into the repository manager plugin  and the repository features of the Plugins Admin Manager to get it to work with Geeklog 2.2.0. It all seems to work for the most part.

It is pretty complex and there is a lot of code clean up to do (along with hard coded html). 

The UI also needs a lot of cleanup (look and functionality). 

Testing is a big thing as well since this is a plugin repository that affects other sites. Making sure that the initial install, upgrades, and patches all work not only from an admin perspective but the webmasters as well. Lots of communicating with the repository server back and forth which makes testing more difficult.

I am not sure if it is worth the time it would take to get it incorporated.

Do you want to take a look? If you do I can branch the code I have on the Geeklog master repository. I would also have to update the Repository Manager with the changes.

Tom

**Comment by mystralkk** [comment]
If a plugin repository means each Geeklog site has its own repository within it and these sites communicate with each other for information of plugins, this sounds like an overkill.  Currently, almost all of our Geeklog plugins are located at GitHub, so GitHub (or BitBucket) should be the only repository.  A repository plugin can just get information and files from the central repository.  So I don't think it is worth while to spend a lot of time incorporating it into Geeklog.

**Comment by eSilverStrike** [comment]
You can define more than one repository (by installing the Repository Manager on other Geeklog sites... I guess a plugin dev might do this) but we would most likely only have one and then the Plugin authors could login into Geeklog.net and upload their plugin or upgrade.

Then webmasters can search for plugins from the Geeklog plugins admin to install or find out if any of the plugins they already have installed have an update.

In a way this is a bit of duplication as authors would have to update the plugin repository with every new release along with also adding it to the Geeklog.net downloads plugin (as you can't download directly from the repository on the Geeklog site for anonymous users, etc..). There is no way for normal users to view the repository except through the plugins admin.

I guess maybe the best option in the future would be to have some sort of combination of a downloads plugin and repository combined.

I'll update the Repository manger with my latest changes. What do you think about the core changes... should I just create a branch for it which we most likely will not use? (just incase someday in the future we do)

**Comment by mystralkk** [comment]
I agree to create a branch for the Repository manager and continue testing it while developing Geeklog core.  After a lot of tests, the Repository manager could be incorporated into the core.

**Comment by eSilverStrike** [comment]
Okay I added the feature/integrate-repository branch for the core code changes so it can work with the updated Repository Manager Plugin https://github.com/Geeklog-GSoC/Repository-Plugin

**Comment by eSilverStrike** [comment]
One more thing to note to test if we implement the repository later: With the repository initially you can upload the plugin via the Repository Manager and specify a version number, what database type (not what min Geeklog Version though) etc... After that you can only upload either patches or upgrades. With upgrades the plugin developer can specify what version the upgrade is for and what it will upgrade too. Most Geeklog plugins don't have specific upgrades released, we just release an install that a user can do a first time install with or can work like an upgrade that also can upgrade any version.

With the repository plugin upgrades must happen in order so I don't think I was using it exactly right. With a new install/upgrade released for a plugin instead of a specific upgrade I think you either must edit the original plugin listed in the Repository (updating the version, file etc) or delete it (along with all of it's old upgrades) and add a new one since a plugin can only have one upgrade tree and patch list.

Patches allow any older version of a plugin to be updated to a specific version. Maybe this is what I should of used for a new plugin repository install. Another thing to look at then.

Anyway the interface for this is a little weird and not intuitive so a lot of testing should be done here to ensure things work correctly and changes may be required to make it more easy to understand to how to use it.

**Comment by hostellerie** [comment]
If this feature is still considered useful, there may be another approach worth looking at before reviving the old Repository Manager code.

I previously developed the **Monitor plugin**, which already included a GitHub-based plugin update mechanism:

https://github.com/Geeklog-Plugins/monitor

Monitor can:

* list installed plugins and their versions;
* query GitHub Releases for supported plugins;
* compare the installed version with the latest available release;
* retrieve the plugin's Geeklog version requirement;
* report whether an update is available and compatible;
* download the new plugin release from GitHub;
* temporarily disable the plugin;
* preserve the previous plugin directories as `.previous`;
* install the downloaded version and then use the normal Geeklog plugin upgrade process.

This was implemented around 2016, so the existing code should not simply be reused unchanged. In particular, the download/update process would need a significant security and compatibility review for Geeklog 2.2.2 and current PHP versions.

For example, a modern implementation should include proper TLS verification, archive validation, protection against path traversal/Zip Slip, compatibility checks, safer staging of files, and ideally a transactional update/rollback mechanism.

The old hard-coded list of supported plugins should also probably be replaced by a more generic catalog or metadata mechanism. GitHub Releases could remain the source of the actual packages, avoiding the need to maintain a separate Geeklog repository server.

One possible architecture would therefore be:

```text
GitHub Releases
      |
      v
Plugin catalog / metadata
      |
      v
Monitor
  - detect installed plugins
  - check available versions
  - report compatibility
      |
      v
Geeklog secure plugin installer/updater
  - download
  - validate
  - stage
  - upgrade
  - rollback if necessary
```

I think the most interesting part of the old Monitor implementation is that it already demonstrated that Geeklog does not necessarily need a dedicated Repository Manager server. GitHub can provide the repository and release infrastructure.

If #822 is still something we want to pursue, I could look at modernizing this part of Monitor for Geeklog 2.2.2 and PHP 8.x and use it as a prototype for a future Plugin Update Manager.

It may also make sense in the longer term for the sensitive installation/update operations to belong to the Geeklog Core, while Monitor only handles discovery, version checking and reporting. This would give plugins a single secure Core API for installation, updates and rollback instead of having each plugin manipulate archives and application directories itself.
